### ISSUE TITLE
Error on encoding after removing the first entry from an array

### ISSUE THREAD

**Issue by neverdane** [remorhaz/php-json-path#25]
Hello! 

I encounter the following exception when I use `$processor->delete` to remove an entry from an array which is not the last: 

```
Invalid element key in decoded JSON: 1 at [path]
```

This exception is triggered by `NodeArrayValue::createChildIterator`:  

```php
public function createChildIterator(): Iterator
{
    $validIndex = 0;
    /** @psalm-var mixed $element */
    foreach ($this->data as $index => $element) {
        if (!is_int($index) || $index != $validIndex++) {
            throw new Exception\InvalidElementKeyException($index, $this->path);
        }
        yield $index => $this
            ->valueFactory
            ->createValue($element, $this->path->copyWithElement($index));
    }
}
```

Here is an example with a failing test:

```php
public function testDelete_FirstArrayEntryQuery_ResultContainsMatchingData(): void
{
    $processor = Processor::create();
    $query = QueryFactory::create()->createQuery('$..a');
    $rootValue = NodeValueFactory::create()->createValue('{"a":1,"b":{"a":2,"c":3}}');

    $result = $processor->delete($query, $rootValue);
    self::assertSame('{"b":{"c":3}}', $result->encode());
}
```

I tried to fix this issue and open a PR by providing a flag to `decode` and `encode` method allowing to reorder arrays but it seems that my fix would be split between your two repositories `remorhaz/php-json-path` and `remorhaz/php-json-data`.

For now, I duplicated your `EventDecoder` and `DecodedJson\NodeValueFactory` in my code base to provide my own `NodeArrayValue` which just ignores the exception when array indexes are not valid (which is not ideal). 

Do you have any idea on how to fix this issue? Perhaps allow with a flag your `delete` method to reorder arrays after a deletion?

Thanks for your time!

**Comment by remorhaz** [comment]
Thank you for your report, I will try to investigate it in nearest time.

**Comment by remorhaz** [comment]
Fixed in 0.8.1.
